### DISCUSSION TITLE
how to temp to switch the repo url when using pdm

### DISCUSSION THREAD

**Discussion by jiangxiaoqiang** [pdm-project/pdm#2406] [Q&A] [answered]
I am using pdm 2.10.1 to manage the python project, now I want to change the download repo url to official when download some new packages. is it possible to specify the download url? what I am doing is add configuration in the pyproject.toml file like this:

```

[[tool.pdm.source]]
url = "https://pypi.org/simple"
name = "pypi"
verify_ssl = true
```

this works. But I found a issue that the download seems did not use the local proxy. it always use the extra source url. Because when I add new packages the local proxy nexus did not download the not exists package. when I remove the config the nexus download works.

so what is the correct way to switch to the official repo when downlaod some new packages? becuase the china mainlan mirror sync the package may take some while. BTW, I have already read the other similiar issue and tried, and facing the issue. is it possible that when download from mirror failed, then try the extra repo url?

this is how I configure the source in `pdf.toml`:

```
[pypi]
verify_ssl = false

[[tool.pdm.source]]
url = "http://admin:fac31388-163a-4843-9f2f-837689809ec3@192.168.50.90:8092/repository/group-pypi/simple"
name = "nexusgroup"
verify_ssl = false

[[tool.pdm.source]]
url = "http://admin:fac31388-163a-4843-9f2f-837689809ec3@192.168.50.90:8092/repository/pypi-proxy/simple"
name = "nexusproxy"
verify_ssl = false

```

and this is the pdm config look like:

```
$ pdm config                                                                                                                                         ‹ruby-2.7.2›
Site/default configuration (/Library/Application Support/pdm/config.toml):
build_isolation = True
cache_dir = /Users/dolphin/Library/Caches/pdm
check_update = True
global_project.fallback = False
global_project.fallback_verbose = True
global_project.path = /Users/dolphin/Library/Application Support/pdm/global-project
global_project.user_site = False
install.cache = False
install.cache_method = symlink
install.parallel = True
pypi.ignore_stored_index = False
pypi.json_api = False
pypi.url = https://pypi.org/simple
pypi.verify_ssl = True
python.providers = ['venv', 'path', 'asdf', 'pyenv', 'rye', 'winreg', 'macos']
python.use_pyenv = True
python.use_venv = True
strategy.resolve_max_rounds = 10000
strategy.save = minimum
strategy.update = reuse
theme.error = red
theme.info = blue
theme.primary = cyan
theme.req = bold green
theme.success = green
theme.warning = yellow
venv.backend = virtualenv
venv.in_project = True
venv.location = /Users/dolphin/Library/Application Support/pdm/venvs
venv.prompt = {project_name}-{python_version}
venv.with_pip = False

Home configuration (/Users/dolphin/Library/Application Support/pdm/config.toml):
check_update = False
pypi.extra.url = https://pypi.org/simple
pypi.nexusgroup.url = http://192.168.50.90:8092/repository/group-pypi/simple
pypi.nexusgroup.username = admin
pypi.nexusgroup.password = <hidden>
pypi.password = <hidden>
pypi.url = http://192.168.50.90:8092/repository/pypi-proxy/simple
pypi.username = admin

Project configuration (/Users/dolphin/source/visa/pdm.toml):
pypi.verify_ssl = False
(venv) (base) 

```

I am using the macOS 13.4.

**Comment by frostming** [comment]
Can you show the full content of all indexes being used, including `tool.pdm.sources` and `pdm config` and what you want to achieve? Because I can't see how you configured the extra source but you mentioned it. You can redact sensitive information.

**Reply by jiangxiaoqiang** [reply]
what I want is that when download the package, first tried the local proxy url http://192.168.50.90:8092/repository/group-pypi/simple, if download failed. try the extra url.  or I can specify the download url when using pdm command for tempura change the default url.

**Reply by frostming** [reply, marked as answer]
try this:

```toml
[tool.pdm.resolution]
respect-source-order = true

[[tool.pdm.source]]
name = "pypi"
url = "http://192.168.50.90:8092/repository/group-pypi/simple"
verify_ssl = true

[[tool.pdm.source]]
name = "fallback"
url = "http://pypi.org/simple"
verify_ssl = true
```
